### PR TITLE
[API] Pick up cart implementation

### PR DESCRIPTION
**Pull request by lchrusciel**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | Related to #11250
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/PropertyInfo/Extractor/EmptyPropertyListExtractor.php [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace Sylius\Bundle\ApiBundle\PropertyInfo\Extractor;
+
+use Symfony\Component\PropertyInfo\PropertyListExtractorInterface;
+
+final class EmptyPropertyListExtractor implements PropertyListExtractorInterface
+{
+    public function getProperties(string $class, array $context = []): ?array
+    {
+        if(class_exists($class)) {
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/PropertyInfo/Extractor/EmptyPropertyListExtractor.php:13]:
```suggestion
        if (class_exists($class)) {
```

File: src/Sylius/Bundle/ApiBundle/spec/PropertyInfo/Extractor/EmptyPropertyListExtractorSpec.php [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    {
+        $this->shouldImplement(PropertyListExtractorInterface::class);
+    }
+
+    function it_provides_empty_list_if_requested_class_exists(): void
+    {
+        $this->getProperties(PickupCart::class, [])->shouldReturn([]);
+    }
+
+    function it_provides_null_if_requested_class_does_not_exists(): void
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/spec/PropertyInfo/Extractor/EmptyPropertyListExtractorSpec.php:32]:
```suggestion
    function it_provides_null_if_requested_class_does_not_exist(): void
```

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    /**
+     * @Then my cart should be empty
+     */
+    public function myCartShouldBeEmpty(): void
+    {
+        $response = $this->cartsClient->getLastResponse();
+        Assert::true(
+            $this->responseChecker->isCreationSuccessful($response),
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:51]: I'm wondering, cannot we _really_ check is there nothing in the created cart? Of course, it needs to be created first, but the content is crucial for this step imo 🐴

File: src/Sylius/Bundle/ApiBundle/CommandHandler/PickupCartHandler.php [hunk 4]
```diff
@@ -0,0 +1,64 @@
+<?php
+
+namespace Sylius\Bundle\ApiBundle\CommandHandler;
```
Comment (Zales0123) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/PickupCartHandler.php:3]: Missing license block 💃

File: src/Sylius/Bundle/ApiBundle/CommandHandler/PickupCartHandler.php [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        $currency = $channel->getBaseCurrency();
+
+        $cart->setChannel($channel);
+        $cart->setLocaleCode($locale->getCode());
+        $cart->setCurrencyCode($currency->getCode());
+        $cart->setTokenValue($this->generator->generateUriSafeString(10));
+
+        $this->orderManager->persist($cart);
+
+        return $cart;
```
Comment (Zales0123) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/PickupCartHandler.php:62]: How orthodox we want to be regarding command and handlers in API? 😄 I theory, the command handler should not return any value

File: src/Sylius/Bundle/ApiBundle/PropertyInfo/Extractor/EmptyPropertyListExtractor.php [hunk 6]
```diff
@@ -0,0 +1,19 @@
+<?php
+
+declare(strict_types=1);
```
Comment (Zales0123) [inline src/Sylius/Bundle/ApiBundle/PropertyInfo/Extractor/EmptyPropertyListExtractor.php:3]: Missing license block 💃

File: src/Sylius/Behat/Client/RequestInterface.php [hunk 7]
```diff
@@ -56,4 +56,6 @@ public function updateFiles(array $newFiles): void;
     public function addSubResource(string $key, array $subResource): void;
 
     public function removeSubResource(string $subResource, string $id): void;
+
+    public function authorize(string $token): void;
```
Comment (Zales0123) [inline src/Sylius/Behat/Client/RequestInterface.php:60]: What is the reason for extracting this method?

File: src/Sylius/Bundle/ApiBundle/CommandHandler/PickupCartHandler.php (hunk 5 above)
Comment (lchrusciel) [reply src/Sylius/Bundle/ApiBundle/CommandHandler/PickupCartHandler.php:62]: There is no other way ATM. At least without a custom data transformer at the end.

File: src/Sylius/Behat/Client/RequestInterface.php (hunk 7 above)
Comment (lchrusciel) [reply src/Sylius/Behat/Client/RequestInterface.php:60]: I need to make a request without authorization, therefore the previous `create` method was no-go for me. However, the rest of the methods still needs to authorize so IMHO it is better to provide it as a part of Public API. BTW, I would like completely decouple Request creation from APIClient and I see these changes as a first steps in this direction.

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php (hunk 3 above)
Comment (lchrusciel) [reply src/Sylius/Behat/Context/Api/Shop/CartContext.php:51]: For now, I would like to leave it as it is due to time constraints

**Comment by GSadee** [comment]
Thanks, Łukasz! :1st_place_medal:

File: src/Sylius/Bundle/ApiBundle/CommandHandler/PickupCartHandler.php (hunk 5 above)
Comment (Zales0123) [reply src/Sylius/Bundle/ApiBundle/CommandHandler/PickupCartHandler.php:70]: So probably we should implement a custom data transformer as fast as possible. The limits our used tool should no be an excuse for the wrong CQS pattern usage 🖖

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php (hunk 3 above)
Comment (Zales0123) [reply src/Sylius/Behat/Context/Api/Shop/CartContext.php:51]: 👍 but still should be changed 🚀

**Pull request merged**